### PR TITLE
Async close of `IndexShard`

### PR DESCRIPTION
**Pull request by DaveCTurner**
Moves the work to close an `IndexShard`, including any final flush and
waiting for merges to complete, off the cluster applier thread to avoid
delaying the application of cluster state updates.

Relates #89821
Relates #107513
Relates #108096
Relates ES-8334

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/ESSingleNodeTestCase.java [hunk 1]
```diff
@@ -460,4 +463,11 @@ protected void ensureNoInitializingShards() {
     protected boolean enableConcurrentSearch() {
         return true;
     }
+
+    protected void awaitIndexShardCloseAsyncTasks() {
+        // ES-8334 TODO build this wait into the relevant APIs (especially, delete-index and close-index)
```
Comment (henningandersen) [inline test/framework/src/main/java/org/elasticsearch/test/ESSingleNodeTestCase.java:468]: I wonder about this, it seems that responding early is fine, given that for instance delete-index still does not guarantee files being deleted?

**Review by henningandersen — COMMENTED** [review]
Looks great. I wonder if we can add an integration test to validate that closing shards no longer blocks the applier thread?

File: server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java [hunk 2]
```diff
@@ -223,8 +219,52 @@ protected void doStop() {
...
+        }
+    }
+
+    /**
+     * @param action Action to run when all the shards removed by earlier-applied cluster states have fully closed. May run on the calling
+     *               thread, or on the thread that completed the closing of the last such shard.
+     */
+    public void onClusterStateShardsClosed(Runnable action) {
+        // In practice, must be called from the applier thread and we should validate the last-applied cluster state version too
+        // ES-8334 TODO validate this is called properly
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java:249]: I've left a few items like this on the JIRA ticket, not intending to solve them here but will follow up in due course.

File: test/framework/src/main/java/org/elasticsearch/test/ESSingleNodeTestCase.java (hunk 1 above)
Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/test/ESSingleNodeTestCase.java:468]: Yeah I'm not sure about this either, you're right that we don't have particularly well-defined guarantees in this area today. We do rely on prompt deletion/closure in some tests tho, enough that it seemed worth a conversation. Again, not intending to solve this here, will follow up.

**Comment by DaveCTurner** [comment]
> I wonder if we can add an integration test to validate that closing shards no longer blocks the applier thread?

Possibly although between `org.elasticsearch.index.shard.IndexShardTests#testAsyncCloseShard` and `org.elasticsearch.cluster.service.ClusterApplierService#assertNotApplyingClusterState` I think this is fairly well covered already. Closing a shard can technically still block the applier thread in the exception-handling code at the bottom of `org.elasticsearch.index.IndexService#createShard`, although there can't be an engine to close at this point (already covered by assertions).

**Review by henningandersen — APPROVED** [review]
LGTM.

**Pull request merged**